### DISCUSSION TITLE
weird behavior in optimization when using symmetries.

### DISCUSSION THREAD

**Discussion by hershey806** [netket/netket#2242] [Q&A]
## The Issue

I am going through the first NetKet tutorial (ground state Ising model). Everything works fine until the section on symmetries. (https://netket.readthedocs.io/en/latest/tutorials/gs-ising.html#translation-symmetry). 

In my first run, the energy just blew up. In later runs, it's either just stuck in the beginning with no samples and initial E =-20.5 or also blows up later. Note that I am not using the same vstate from before, I rerun it and also tried restarting the kernel and only running the symmetries part without any of the previous parts. This is specifically what I am running then:
```
N = 20
hi = nk.hilbert.Spin(s=1 / 2, N=N)
graph = nk.graph.Chain(length=N, pbc=True)

Gamma = -1
H = sum([Gamma * sigmax(hi, i) for i in range(N)])

V = -1
H += sum([V * sigmaz(hi, i) @ sigmaz(hi, j) for (i, j) in graph.edges()])

import flax.linen as nn
import netket.nn as nknn
from flax import nnx
import jax.numpy as jnp

class SymmModel(nnx.Module):

    def __init__(self, N: int, alpha: int = 1, *, rngs: nnx.Rngs):
        """
        Construct a Feed-Forward Neural Network with a single hidden layer.

        Args:
            N: The number of input nodes (number of spins in the chain).
            alpha: The density of the hidden layer. The hidden layer will have
                N*alpha nodes.
            rngs: The random number generator seed.
        """
        self.alpha = alpha

        # We want to use netket's DenseSymm layer, which requires a symmetry group.
        dense_symm_linen = nknn.DenseSymm(
            symmetries=graph.translation_group(),
            features=alpha,
            kernel_init=nn.initializers.normal(stddev=0.01),
        )
        # However, this layer is defined using ``flax.linen``, so to use it with NNX we
        # must resort to the conversion function ``nnx.bridge.ToNNX``.
        # The syntax is a bit weird, as we need to separately pass the random number generator
        # and then initialize the layer with a dummy input of shape (1, 1, Nsites).
        self.linear_symm = nnx.bridge.ToNNX(dense_symm_linen, rngs=rngs).lazy_init(
            jnp.ones((1, 1, N))
        )

    def __call__(self, x: jax.Array):
        # add an extra dimension with size 1, because DenseSymm requires rank-3 tensors as inputs.
        # the shape will now be (batches, 1, Nsites)
        x = x.reshape(-1, 1, x.shape[-1])

        x = self.linear_symm(x)
        x = nnx.relu(x)

        # sum the output
        return jnp.sum(x, axis=(-1, -2))


sampler = nk.sampler.MetropolisLocal(hi)

# Let us define a model with 4 features per symmetry
model = SymmModel(N=N, alpha=4, rngs=nnx.Rngs(2))

vstate = nk.vqs.MCState(sampler, model, n_samples=1008)

vstate.n_parameters

optimizer = nk.optimizer.Sgd(learning_rate=0.1)

gs = nk.driver.VMC(
    H,
    optimizer,
    variational_state=vstate,
    preconditioner=nk.optimizer.SR(diag_shift=0.1), 
)

log = nk.logging.RuntimeLog()
gs.run(n_iter=600, out=log)

symm_energy = vstate.expect(H)
error = abs((symm_energy.mean - eig_vals[0]) / eig_vals[0])
print("Optimized energy and relative error: ", symm_energy, error)
```

## The Output

Just stuck on this: `   0%|          | 0/600 [00:06<?, ?it/s, Energy=-20.59 ± 0.15 [σ²=1.8e+01, R̂=1.044]]
`

Or energy increases and plateaus or just blows up as I mentioned before (though it is more commonly just stuck). 

## Environment

I am running on NERSC Perlmutter Jupyter on a GPU node with 4 NVIDIA A100 GPU's. I'm using the latest (unreleased) NetKet verison because I was getting JAX errors otherwise. Here are all the versions:
```
Python version:  3.11.7
NetKet version:  3.22.0.dev4
JAX version:     0.10.1
jaxlib version:  0.10.1
Flax version:    0.12.7
Plum version:    2.9.0
```

**Comment by PhilipVinc** [comment]
If I run the example above I get
```python
100%|████████████████████████████████████| 600/600 [00:19<00:00, 31.03it/s, Energy=-25.4879 ± 0.0018 [σ²=3.4e-03, R̂=1.006]]]
```

so I think it works?

**Comment by hershey806** [comment]
It's probably just some GPU thing... it works on CPU but I'm not sure why.

**Comment by PhilipVinc** [comment]
it's weird, it should not change result in gpu vs cpu. though the lower numerical accuracy of some gpu algorithms might be to blame. You could try to swap to 'pinv_smooth' solver for the linear system which is more stable, or lower the learning rate that is quite aggressive and might lead to instabilities...
